www/js/lib/xzdec_wrapper.js

```diff
@@ -20,11 +20,56 @@
  * along with Kiwix (file LICENSE-GPLv3.txt).  If not, see <http://www.gnu.org/licenses/>
  */
 'use strict';
-define(['xzdec'], function() {
-    // DEV: xzdec.js emits a global Module variable, which cannot be set in requireJS function line above, though it can be loaded in definition
-    var xzdec = Module;
-    xzdec._init();
-    
+
+// DEV: Put your RequireJS definition in the rqDefXZ array below, and any function exports in the function parenthesis of the define statement
+// We need to do it this way in order to load the wasm or asm versions of xzdec conditionally. Older browsers can only use the asm version
+// because they cannot interpret WebAssembly.
+var rqDefXZ = [];
+
+// Select asm or wasm conditionally
+if ('WebAssembly' in self) {
+    console.debug('Using WASM xz decoder');
+    rqDefXZ.push('xzdec-wasm');
+} else {
+    console.debug('Using ASM xz decoder');
+    rqDefXZ.push('xzdec-asm');
+}
+
+define(rqDefXZ, function() {
+    // DEV: xzdec.js has been compiled with `-s EXPORT_NAME="XZ" -s MODULARIZE=1` to avoid a clash with zstddec.js
+    // Note that we include xzdec-asm or xzdec-wasm above in requireJS definition, but we cannot change the name in the function list
+    // There is no longer any need to load it in index.html
+    // For explanation of loading method below to avoid conflicts, see https://github.com/emscripten-core/emscripten/blob/master/src/settings.js
+
+    /**
+     * @typedef EMSInstance An object type representing an Emscripten instance
+     */
+
+    /**
+     * The XZ Decoder instance
+     * @type EMSInstance
+     */
+     var xzdec;
+
+     var instantiateDecoder = function (instance) {
+         xzdec = instance;
+     };
+
+     XZ().then(instantiateDecoder)
+     .catch(function (err) {
+         console.debug(err);
+         if (/CompileError.+?WASM/i.test(err.message)) {
```

Comment:
I'm wondering is we should not always try to load the ASM version if the WASM fails, regardless of the error message (that could be different, depending on the browser).
If not, the log level should be higher IMHO when the error message does not match

Reply:
OK, yes you're right, we don't need to know exactly why instantiation failed, just that it failed.